src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+                for (int i = 0; i < length; ++i)
+                {
+                    _magList[i] = Math.Sqrt((Math.Pow(_fftRe[i], 2) + Math.Pow(_fftIm[i], 2)));
+                    if (_magList[i] > _eps)
+                    {
+                        _magLogList[i] = Math.Log(_magList[i]);
+                    }
+                    else
+                    {
+                        _magLogList[i] = 0;
+                    }
+                }
+
+                // Step 4: Calculate spectral
+                double[] filteredLogList = AverageFilter(_magLogList, _averagingWindowSize);
+                AllocateDoubleArray(ref _spectralList, length);
+                for (int i = 0; i < length; ++i)
+                {
+                    _spectralList[i] = Math.Exp(_magLogList[i] - filteredLogList[i]);
+                }
+
+                // Step 5: IFFT transformation
+                AllocateDoubleArray(ref _transRe, length);
+                AllocateDoubleArray(ref _transIm, length);
+                for (int i = 0; i < length; ++i)
+                {
+                    if (_magLogList[i] != 0)
+                    {
+                        _transRe[i] = _fftRe[i] * _spectralList[i] / _magList[i];
+                        _transIm[i] = _fftIm[i] * _spectralList[i] / _magList[i];
+                    }
+                    else
+                    {
+                        _transRe[i] = 0;
+                        _transIm[i] = 0;
+                    }
+                }
+
+                AllocateDoubleArray(ref _ifftRe, length);
+                AllocateDoubleArray(ref _ifftIm, length);
+                FftUtils.ComputeBackwardFft(_transRe, _transIm, _ifftRe, _ifftIm, length);
+
+                // Step 6: Calculate mag and ave_mag of IFFT
+                AllocateDoubleArray(ref _ifftMagList, length);
+                for (int i = 0; i < length; ++i)
+                {
+                    _ifftMagList[i] = Math.Sqrt((Math.Pow(_ifftRe[i], 2) + Math.Pow(_ifftIm[i], 2)));
+                }
+                double[] filteredIfftMagList = AverageFilter(_ifftMagList, Math.Min(_ifftMagList.Length, _judgementWindowSize));
+
+                // Step 7: Calculate raw score and set result
+                for (int i = 0; i < results.GetLength(0); ++i)
+                {
+                    var score = CalculateScore(_ifftMagList[i], filteredIfftMagList[i]);
+                    score /= 10.0f;
+                    score = Math.Min(score, 1);
+                    score = Math.Max(score, 0);
+
+                    var detres = score > threshold ? 1 : 0;
+
+                    results[i][0] = detres;
+                    results[i][1] = score;
+                    results[i][2] = _ifftMagList[i];
+                }
+            }
+
+            private double[] BackAdd(double[] data)
+            {
+                AllocateDoubleArray(ref _predictArray, _lookaheadWindowSize + 1);
```

Comment:
>_predictArray [](start = 40, length = 13)

This is always allocated to be the same size, and doesn't depend on the batch size. Can be readonly and instantiated in the constructor.

Reply:
done.

---
In reply to: [427853277](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427853277) [](ancestors = 427853277)